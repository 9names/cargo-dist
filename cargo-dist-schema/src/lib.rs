use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
// FIXME: ideally these would be UTF8PathBufs but JsonSchema doesn't support (yet?)
use std::path::PathBuf;

/// The final report of cargo-dist
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct DistReport {
    /// App releases we're distributing
    pub releases: Vec<Release>,
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Release {
    /// The name of the app
    pub app_name: String,
    /// The version of the app
    // FIXME: should be a Version but JsonSchema doesn't support (yet?)
    pub app_version: String,
    /// The distributable bundles for this release
    pub distributables: Vec<Distributable>,
}

/// A distributable bundle that's part of a Release
///
/// i.e. a zip or installer
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct Distributable {
    /// The kind of distributable (e.g. "zip")
    pub kind: DistributableKind,
    /// The target triple of the bundle
    pub target_triple: String,
    /// The location of the distributable bundle
    pub path: PathBuf,
    /// Artifacts included in the bundle (like executables)
    pub artifacts: Vec<Artifact>,
}

/// An artifact included in a Distributable
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
#[serde(tag = "kind")]
pub enum Artifact {
    /// An executable artifact
    #[serde(rename = "executable")]
    Executable(ExecutableArtifact),
}

#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub enum DistributableKind {
    /// A zip or a tarball
    #[serde(rename = "zip")]
    Zip,
}

/// An executable artifact (exe/binary)
#[derive(Debug, Clone, Serialize, Deserialize, JsonSchema)]
pub struct ExecutableArtifact {
    /// The name of the executable
    pub name: String,
    /// The path of the executable relative to the root of the package
    pub path: PathBuf,
}

impl DistReport {
    pub fn new(releases: Vec<Release>) -> Self {
        Self { releases }
    }
}

#[test]
fn emit() {
    let schema = schemars::schema_for!(DistReport);
    insta::assert_snapshot!(serde_json::to_string_pretty(&schema).unwrap());
}
