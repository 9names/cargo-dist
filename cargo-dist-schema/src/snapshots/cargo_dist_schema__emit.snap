---
source: cargo-dist-schema/src/lib.rs
expression: "serde_json::to_string_pretty(&schema).unwrap()"
---
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "title": "DistReport",
  "description": "The final report of cargo-dist",
  "type": "object",
  "required": [
    "releases"
  ],
  "properties": {
    "releases": {
      "description": "App releases we're distributing",
      "type": "array",
      "items": {
        "$ref": "#/definitions/Release"
      }
    }
  },
  "definitions": {
    "Artifact": {
      "description": "An artifact included in a Distributable",
      "oneOf": [
        {
          "description": "An executable artifact",
          "type": "object",
          "required": [
            "kind",
            "name",
            "path"
          ],
          "properties": {
            "kind": {
              "type": "string",
              "enum": [
                "executable"
              ]
            },
            "name": {
              "description": "The name of the executable",
              "type": "string"
            },
            "path": {
              "description": "The path of the executable relative to the root of the package",
              "type": "string"
            }
          }
        }
      ]
    },
    "Distributable": {
      "description": "A distributable bundle that's part of a Release\n\ni.e. a zip or installer",
      "type": "object",
      "required": [
        "artifacts",
        "kind",
        "path",
        "target_triple"
      ],
      "properties": {
        "artifacts": {
          "description": "Artifacts included in the bundle (like executables)",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Artifact"
          }
        },
        "kind": {
          "description": "The kind of distributable (e.g. \"zip\")",
          "allOf": [
            {
              "$ref": "#/definitions/DistributableKind"
            }
          ]
        },
        "path": {
          "description": "The location of the distributable bundle",
          "type": "string"
        },
        "target_triple": {
          "description": "The target triple of the bundle",
          "type": "string"
        }
      }
    },
    "DistributableKind": {
      "oneOf": [
        {
          "description": "A zip or a tarball",
          "type": "string",
          "enum": [
            "zip"
          ]
        }
      ]
    },
    "Release": {
      "type": "object",
      "required": [
        "app_name",
        "app_version",
        "distributables"
      ],
      "properties": {
        "app_name": {
          "description": "The name of the app",
          "type": "string"
        },
        "app_version": {
          "description": "The version of the app",
          "type": "string"
        },
        "distributables": {
          "description": "The distributable bundles for this release",
          "type": "array",
          "items": {
            "$ref": "#/definitions/Distributable"
          }
        }
      }
    }
  }
}
